File: OneDrive/Artificial Intelligence Learn/Anil_App/backend/app/models/analysis.py
"""Pydantic models for unified stock analysis response."""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict


class TechnicalSnapshot(BaseModel):
    """AI-estimated technical indicator values."""

    sma_20: float | None = None
    sma_50: float | None = None
    sma_200: float | None = None
    ema_12: float | None = None
    ema_26: float | None = None
    rsi_14: float | None = None
    macd_line: float | None = None
    macd_signal: float | None = None
    macd_histogram: float | None = None
    bollinger_upper: float | None = None
    bollinger_middle: float | None = None
    bollinger_lower: float | None = None
    support_levels: list[float] = []
    resistance_levels: list[float] = []
    signal: Literal["strong_buy", "buy", "neutral", "sell", "strong_sell"] = "neutral"


class NewsItem(BaseModel):
    """A single news headline from AI analysis."""

    title: str
    source: str | None = None
    sentiment: Literal["positive", "negative", "neutral"] | None = None


class QuarterlyEarning(BaseModel):
    """One quarter of earnings data."""

    quarter: str
    revenue: float | None = None
    net_income: float | None = None
    eps: float | None = None
    yoy_revenue_growth: float | None = None


class PriceForecast(BaseModel):
    """Low / mid / high price forecast with a confidence score."""

    low: float
    mid: float
    high: float
    confidence: float


class PricePredictions(BaseModel):
    """Forecasts across three time horizons."""

    one_week: PriceForecast
    one_month: PriceForecast
    three_months: PriceForecast


class RiskAssessment(BaseModel):
    """Qualitative and quantitative risk evaluation."""

    overall_risk: Literal["low", "medium", "high", "very_high"]
    risk_factors: list[str]
    risk_score: float


class LongTermOutlook(BaseModel):
    """Long-term investment outlook with multi-year projections."""

    one_year: PriceForecast
    five_year: PriceForecast
    ten_year: PriceForecast
    verdict: Literal["strong_buy", "buy", "hold", "sell", "strong_sell"]
    verdict_rationale: str
    catalysts: list[str] = []
    long_term_risks: list[str] = []
    compound_annual_return: float


class HistoricalPrice(BaseModel):
    """Single day OHLC price point for charting."""

    date: str
    open: float
    high: float
    low: float
    close: float
    volume: int | None = None


class StockAnalysisResponse(BaseModel):
    """Full unified stock analysis — everything from one AI call."""

    model_config = ConfigDict(protected_namespaces=())

    # Quote data
    ticker: str
    company_name: str
    current_price: float
    previous_close: float | None = None
    open: float | None = None
    day_high: float | None = None
    day_low: float | None = None
    volume: int | None = None
    # Fundamentals
    market_cap: str | None = None
    pe_ratio: float | None = None
    eps: float | None = None
    week_52_high: float | None = None
    week_52_low: float | None = None
    dividend_yield: float | None = None
    # Technical
    technical: TechnicalSnapshot | None = None
    # News
    news: list[NewsItem] = []
    quarterly_earnings: list[QuarterlyEarning] = []
    # Historical
    historical_prices: list[HistoricalPrice] = []
    # Company info
    company_description: str = ""
    sector: str = ""
    industry: str = ""
    headquarters: str = ""
    ceo: str = ""
    founded: str = ""
    employees: str = ""
    # AI Analysis
    recommendation: Literal["strong_buy", "buy", "hold", "sell", "strong_sell"]
    confidence_score: float
    summary: str
    bull_case: str
    bear_case: str
    risk_assessment: RiskAssessment
    price_predictions: PricePredictions
    long_term_outlook: LongTermOutlook | None = None
    # Research enrichment
    research_context: str = ""
    research_sources: list[str] = []
    # Metadata
    analysis_timestamp: datetime
    model_used: str = "gpt-5.3"
    disclaimer: str = (
        "Price data sourced from Financial Modeling Prep. "
        "Analysis and predictions are AI-generated — not financial advice."
    )

File: OneDrive/Artificial Intelligence Learn/Anil_App/frontend/src/types/analysis.ts
export interface SearchResult {
  symbol: string;
  name: string;
}

export type Recommendation = 'strong_buy' | 'buy' | 'hold' | 'sell' | 'strong_sell';
export type RiskLevel = 'low' | 'medium' | 'high' | 'very_high';
export type TechnicalSignal = 'strong_buy' | 'buy' | 'neutral' | 'sell' | 'strong_sell';
export type Sentiment = 'positive' | 'negative' | 'neutral';

export interface TechnicalSnapshot {
  sma_20: number | null;
  sma_50: number | null;
  sma_200: number | null;
  ema_12: number | null;
  ema_26: number | null;
  rsi_14: number | null;
  macd_line: number | null;
  macd_signal: number | null;
  macd_histogram: number | null;
  bollinger_upper: number | null;
  bollinger_middle: number | null;
  bollinger_lower: number | null;
  support_levels: number[];
  resistance_levels: number[];
  signal: TechnicalSignal;
}

export interface NewsItem {
  title: string;
  source: string | null;
  sentiment: Sentiment | null;
}

export interface QuarterlyEarning {
  quarter: string;
  revenue: number | null;
  net_income: number | null;
  eps: number | null;
  yoy_revenue_growth: number | null;
}

export interface HistoricalPrice {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number | null;
}

export type AnalysisTab = 'chart' | 'news' | 'financials' | 'about' | 'invest';

export interface PriceForecast {
  low: number;
  mid: number;
  high: number;
  confidence: number;
}

export interface PricePredictions {
  one_week: PriceForecast;
  one_month: PriceForecast;
  three_months: PriceForecast;
}

export interface LongTermOutlook {
  one_year: PriceForecast;
  five_year: PriceForecast;
  ten_year: PriceForecast;
  verdict: Recommendation;
  verdict_rationale: string;
  catalysts: string[];
  long_term_risks: string[];
  compound_annual_return: number;
}

export interface RiskAssessment {
  overall_risk: RiskLevel;
  risk_factors: string[];
  risk_score: number;
}

export interface StockAnalysisResponse {
  ticker: string;
  company_name: string;
  current_price: number;
  previous_close: number | null;
  open: number | null;
  day_high: number | null;
  day_low: number | null;
  volume: number | null;
  market_cap: string | null;
  pe_ratio: number | null;
  eps: number | null;
  week_52_high: number | null;
  week_52_low: number | null;
  dividend_yield: number | null;
  technical: TechnicalSnapshot | null;
  news: NewsItem[];
  quarterly_earnings: QuarterlyEarning[];
  historical_prices: HistoricalPrice[];
  company_description: string;
  sector: string;
  industry: string;
  headquarters: string;
  ceo: string;
  founded: string;
  employees: string;
  recommendation: Recommendation;
  confidence_score: number;
  summary: string;
  bull_case: string;
  bear_case: string;
  risk_assessment: RiskAssessment;
  price_predictions: PricePredictions;
  long_term_outlook: LongTermOutlook | null;
  research_context: string;
  research_sources: string[];
  analysis_timestamp: string;
  model_used: string;
  disclaimer: string;
}
